test/Mono.Linker.Tests.Cases/Basic/LinkerHandlesRefFields.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+using Mono.Linker.Tests.Cases.Expectations.Assertions;
+
+namespace Mono.Linker.Tests.Cases.Basic
+{
+	[ExpectedNoWarnings]
+	[SkipKeptItemsValidation]
+	class LinkerHandlesRefFields
+	{
+		ref struct RS
+		{
+			public ref int i;
+			public ref double d;
+			public RS2 rs2;
+			public ref string str;
+			public ref object obj;
+			public ref Type t;
+			public void MoveThis ()
+			{
+				this = new RS ();
+			}
+		}
+
+		ref struct RS2
+		{
+			public ref int i;
+			public ref double d;
+			public ref string str;
+			public ref object obj;
+			public ref Type t;
+			public void MoveThis ()
+			{
+				this = new RS2 ();
+			}
+		}
+
+		delegate RS2 RsParamRs2Return (RS rs);
+		delegate ref RS2 RefRSParamRefRs2Return (ref RS rs);
+		delegate ref int RsParamRefIntReturn (RS rs);
+
+		public static void Main (string[] args)
```

Comment:
I actually don't know if some of these are valid, but still:
- ref struct in a generic parameter
- ref return and parameter on a virtual method?
- Put annotations (DAM/RUC on type) on the ref field - don't care if it works correctly, but should not crash

Reply:
> ref struct in a generic parameter
Ref structs will generate [CS0306](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0306) if used as a type parameter.

> ref return and parameter on a virtual method?
Added IRS interface and Base and Derived classes with virtual methods with ref structs as return, ref return, parameter, and ref parameter.

> Put annotations (DAM/RUC on type) on the ref field
RUC can't be placed on structs and ref structs can't implement interfaces, so I don't think there's a way that they could be annotated with DAM. I've added DAM on a field and DAM and RUC on a method return and parameter.

Reply:
Thanks a lot!